src/hammer-vlsi/hammer_vlsi/constraints.py

```diff
@@ -80,19 +82,153 @@ def from_setting(d: dict) -> "SRAMParameters":
+
+        location = None  # type: Optional[Tuple[float, float]]
+        if "location" in raw_assign:
+            location_raw = raw_assign["location"]  # type: Union[List[float], Tuple[float, float]]
+            assert len(location_raw) == 2, "location must be a Optional[Tuple[float, float]]"
+            assert isinstance(location_raw[0], float), "location must be a Optional[Tuple[float, float]]"
+            assert isinstance(location_raw[1], float), "location must be a Optional[Tuple[float, float]]"
+            location = (location_raw[0], location_raw[1])
+        if not semi_auto and location is not None:
+            raise PinAssignmentSemiAutoError("location requires semi_auto")
+
+        width = raw_assign.get("width", None)  # type: Optional[float]
+        if width is not None:
+            assert isinstance(width, float), "width must be a float"
+        if not semi_auto and width is not None:
+            raise PinAssignmentSemiAutoError("width requires semi_auto")
+
+        depth = raw_assign.get("depth", None)  # type: Optional[float]
+        if depth is not None:
+            assert isinstance(depth, float), "depth must be a float"
+        if not semi_auto and depth is not None:
+            raise PinAssignmentSemiAutoError("depth requires semi_auto")
+
+        layers = None  # type: Optional[List[str]]
+        if "layers" in raw_assign:
+            raw_layers = raw_assign["layers"]  # type: List[str]
+            assert isinstance(raw_layers, list), "layers must be a List[str]"
+            for layer in "layers":
+                assert isinstance(layer, str), "layers must be a List[str]"
+            layers = raw_layers
+
+        if preplaced:
+            should_be_none = reduce(operator.and_, map(lambda x: x is None, [side, location, width, depth]))
+            if len(get_or_else(layers, cast(List[str], []))) > 0 or not should_be_none:
+                raise PinAssignmentPreplacedError(
+                    PinAssignment(pins=pins, side=None, layers=[], preplaced=preplaced, location=None, width=None,
+                                  depth=None))
+        else:
+            if len(get_or_else(layers, cast(List[str], []))) == 0 or side is None:
+                raise PinAssignmentError(
+                    "Pins {p} assigned without layers or side. Assuming pins will be handled by CAD tool.".format(
+                        p=pins))
+        return PinAssignment(
+            pins=pins,
+            side=side,
+            layers=layers,
+            preplaced=preplaced if "preplaced" in raw_assign else None,
+            location=location,
+            width=width,
+            depth=depth
+        )
+
+    def to_dict(self) -> dict:
+        base = {
+            "pins": self.pins
+        }  # type: Dict[str, Any]
+        if self.side is not None:
+            base['side'] = self.side
+        if self.layers is not None:
+            base['layers'] = self.layers
+        if self.preplaced is not None:
```

Comment:
As above default should be `Some(False)`

Reply:
Ditto above